intermediate_source/custom_function_conv_bn_tutorial.py

```diff
@@ -0,0 +1,344 @@
+by eliminating the batch norm layer entirely and updating the weight
+and bias of the preceding convolution [0]. This technique is not applicable
+for training models, however.
+
+In this tutorial, we will show a different technique to fuse the two layers
+that can be applied during training. Rather than improved runtime, the
+objective of this optimization is to reduce memory usage.
+
+The idea behind this optimization is to see that both convolution and
+batch norm (as well as many other ops) need to save a copy of their input
+during forward for the backward pass. For large
+batch sizes, these saved inputs are responsible for much of your memory usage,
+so being able to avoid allocating another input tensor for every
+convolution batch norm pair can be a significant reduction.
+
+In this tutorial, we avoid this extra allocation by combining convolution
+and batch norm into a single layer (as a custom function). In the forward
+of this combined layer, we perform normal convolution and batch norm as-is,
+with the only difference being that we will only save the inputs to the convolution.
+To obtain the input of batch norm, which is necessary to backward through
+it, we recompute convolution forward again during the backward pass.
+
+For simplicity, in this tutorial we will not support the bias, stride, dilation, or
+group parameters for convolutions.
+
+[0] https://nenadmarkus.com/p/fusing-batchnorm-and-conv/
+"""
+
+######################################################################
+# Implementing a custom function requires us to implement the backward
+# ourselves. In this case, we need both the backward formulas for Conv2D
+# and BatchNorm2D. Eventually we'd chain them together in our unified
+# backward function, but below we first implement them as their own
+# custom functions so we can validate their correctness individually
+import torch
+from torch.autograd.function import once_differentiable
+import torch.nn.functional as F
+
+class Conv2D(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, X, weight):
+        ctx.save_for_backward(X, weight)
+        return F.conv2d(X, weight)
+
+    # Use @once_differentiable by default unless we intend to double
+    # backward
+    @staticmethod
+    @once_differentiable
+    def backward(ctx, grad_out):
+        X, weight = ctx.saved_tensors
+        grad_input = F.conv2d(X.transpose(0, 1), grad_out.transpose(0, 1)).transpose(0, 1)
+        grad_X = F.conv_transpose2d(grad_out, weight)
+        return grad_X, grad_input
+
+######################################################################
+# When testing with gradcheck, it is important to use double precision
+weight = torch.rand(5, 3, 3, 3, requires_grad=True, dtype=torch.double)
+X = torch.rand(10, 3, 7, 7, requires_grad=True, dtype=torch.double)
+torch.autograd.gradcheck(Conv2D.apply, (X, weight))
+
+######################################################################
+# Batch Norm
+def expand(t):
```

Comment:
The name is not the best here as there is already a torch.expand function that does something else